### PR TITLE
Viscous Residual Prediction: learn delta_p = p_CFD - p_panel

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Instead of predicting the full pressure field, predict only the **viscous correction** `delta_p = p_CFD - p_panel` where `p_panel` is the inviscid thin-airfoil panel Cp (already computed via `--cp_panel`). The model learns only the viscous residual (boundary layer, separation, wake viscous effects) rather than the full pressure field.

**Why this should work:**
- The panel Cp is already a first-order approximation of the pressure. The CFD pressure differs mainly through localized viscous corrections — smaller magnitude, smoother, more structured
- Predicting a smaller residual is an easier regression problem (multigrid preconditioning analogy; image super-resolution residual learning)
- `--residual_prediction` already does this for velocity (subtracts freestream) — this extends the same pattern to pressure using physics-grounded panel Cp baseline
- Reference: arXiv:2205.09025 "Residual-based physics-informed machine learning"

**Key:** `--cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1` control the *input feature*. The new `--viscous_residual_target` changes the *prediction target*. Both can be active simultaneously.

## Instructions

Add `--viscous_residual_target` flag. When active, subtract panel Cp from the pressure *target* (in physical Cp space, before asinh/normalization), for surface nodes only.

**Step 1 — Add flag to `Config` dataclass (around line 1255):**
```python
viscous_residual_target: bool = False  # predict viscous correction delta_p = p_CFD - p_panel
```

**Step 2 — Training loop target modification** (insert BEFORE the `if cfg.asinh_pressure:` block, around line 1938):

```python
if cfg.viscous_residual_target and cfg.cp_panel:
    # Compute raw panel Cp in physical Cp space (unscaled — full strength for target)
    _cp_panel_tgt = compute_cp_panel(_raw_xy_te, _raw_aoa, is_surface, _raw_saf_norm_te)
    if cfg.cp_panel_tandem_only:
        _cp_panel_tgt = _cp_panel_tgt * is_tandem.float()[:, None, None]
    # Subtract from pressure channel (index 2) for surface nodes only
    y_phys[:, :, 2:3] = y_phys[:, :, 2:3] - _cp_panel_tgt * is_surface.float().unsqueeze(-1)
```

Note: `y_phys` must be cloned before modification if it's shared: add `y_phys = y_phys.clone()` before the block. After this, the asinh transform and normalization proceed as normal on the modified `y_phys`.

**Step 3 — Eval/verify loop inverse transform** (apply in both eval and verify loops, after predictions are obtained and before metric computation):

In each eval loop, the targets `y_norm` were built from `y_phys` (with the residual subtracted). Metrics must be computed in original pressure space. After converting predictions back to physical space, add panel Cp back:

```python
if cfg.viscous_residual_target and cfg.cp_panel:
    _cp_panel_tgt_v = compute_cp_panel(_raw_xy_te, _raw_aoa, is_surface, _raw_saf_norm_te)
    if cfg.cp_panel_tandem_only:
        _cp_panel_tgt_v = _cp_panel_tgt_v * is_tandem.float()[:, None, None]
    surf_mask = is_surface.float().unsqueeze(-1)
    # Add cp_panel back to BOTH predictions and targets before metrics
    # In normalized space: cp_panel_norm = cp_panel / y_std[2]
    _y_std_p = phys_stats["y_std"][:, :, 2:3]  # [1, 1, 1] broadcast
    _cp_norm = _cp_panel_tgt_v / _y_std_p
    pred[:, :, 2:3] += _cp_norm * surf_mask      # add back to predictions
    y_norm[:, :, 2:3] += _cp_norm * surf_mask    # add back to targets
```

Apply this in ALL three loops: train validation step, eval loop, and verify loop — anywhere metrics are computed.

**Step 4 — Run 2 seeds:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/viscous-resid-s42" --wandb_group viscous-residual-target \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --viscous_residual_target

CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent edward --wandb_name "edward/viscous-resid-s73" --wandb_group viscous-residual-target \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --viscous_residual_target
```

**Note on asinh interaction:** `--asinh_pressure` transforms the pressure target AFTER our subtraction. The residual (viscous correction) will be asinh-transformed, which is fine — the dynamic range of the residual is smaller than the full pressure, so asinh compression may help even more.

## Baseline (PR #2350, 2-seed avg)

| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.90 | 7.35 | 27.20 | 6.40 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature`

## Round 40 Bold Context

This is the **#1 priority bold experiment** — a fundamental reformulation of the prediction task. The panel Cp is already proven valuable as an input feature (merged, -2.6% p_oodc). Using it as the prediction baseline is the natural next step. Expected primary gains: **p_tan** and **p_in** (the model only needs to learn the viscous correction, not the full inviscid+viscous field).

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: **Baseline updated.** PR #2357 (Vortex-Panel Induced Velocity) was just merged. Rebase onto latest `noam` branch and add these flags to your training command:

```
--vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

New baseline metrics to beat (2-seed avg):
| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.872 | 7.459 | 26.319 | 6.229 |

**Comment by tcapelle** [comment]
STUDENT edward:

## Results (initial run — pre-rebase, incomplete S73)

### Implementation note

During this experiment I discovered a bug in the implementation: the val loop's y_norm construction did not subtract the panel_cp residual before computing the val_loss (unlike the training loop which did). This meant val_loss for tandem/OOD splits was computed as the error between residual predictions and original (non-residual) targets → inflated losses for those splits.

**S42 (hiu2225x)**: Correct implementation — started after the bug fix. Results valid.  
**S73 (cgjvmyrk)**: Ran with buggy val loop (process survived my kill command). S73's checkpoint selection was based on corrupted val_loss for tandem/OOD splits; S73 results are unreliable.

### S42 metrics (correct implementation)

| Metric | S42 (hiu2225x) | Baseline (PR #2350) | Δ |
|--------|----------------|---------------------|---|
| **p_in** | 12.59 | 11.90 | +5.8% ❌ |
| **p_tan** | 27.3 | 27.20 | +0.4% ❌ |
| **p_oodc** | 9.1 | 7.35 | +23.8% ❌ |
| **p_re** | 7.1 | 6.40 | +10.9% ❌ |

Peak memory: 40.6 GB.

### Command used (initial run)

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/viscous-resid-s42" --wandb_group viscous-residual-target \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --viscous_residual_target
```

### Analysis (initial run)

The hypothesis failed. All metrics are significantly worse, especially p_oodc (+23.8%). Root cause:

The flat-plate panel_cp is a geometry-agnostic approximation that works reasonably for in-distribution cases but is a poor baseline for tandem/OOD conditions (where wake interference, extreme AoA, and geometry variations make the flat-plate assumption wrong). Subtracting a poor baseline from the target can make the residual MORE variable and harder to predict, not less — the opposite of the multigrid preconditioning analogy. 

Additionally, the phys_stats normalization was computed from original pressures, not residuals, so the normalized residual targets have a non-standard distribution.

---

## Rebasing onto latest noam (PR #2357 baseline)

Rebasing as requested with  and new baseline metrics. Will post updated results shortly.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2362 because ADVISOR: Closing — Viscous Residual Prediction is a dead end.

**Results (S42, correct implementation):**
| Metric | S42 | Baseline (#2357) | Δ |
|--------|-----|----------|---|
| p_in | 12.59 | 11.872 | +6.0% ❌ |
| p_oodc | 9.1 | 7.459 | +22.0% ❌ |
| p_tan | 27.3 | 26.319 | +3.7% ❌ |
| p_re | 7.1 | 6.229 | +14.0% ❌ |

**Root cause:** Flat-plate panel Cp is too inaccurate to serve as a residual baseline, especially for tandem/OOD conditions where wake interference makes the inviscid estimate wildly wrong. Subtracting an inaccurate baseline makes the residual target MORE variable and harder to predict — the opposite of the intended effect. The phys_stats normalization mismatch (computed from raw pressure, not residuals) compounds the problem.

**Lesson learned:** Residual prediction against a physics baseline requires that baseline to be accurate enough that the residual is simpler than the original target. Panel-method Cp does not meet this bar for tandem foils. Future residual approaches should use a more accurate baseline (e.g., RANS or the model's own first-pass prediction).

Good bug catch on the val loop panel_cp subtraction — that was solid debugging. New assignment incoming..
